### PR TITLE
[TASK] Add Named Arguments best practices

### PR DESCRIPTION
**Pull request by lolli42**
The TYPO3 core team decided on this document
regarding best practices for the adoption and
avoidance of this PHP language feature.

### CODE REVIEW

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:131]:
```suggestion
* TYPO3 Core tries to be as consistent as possible
```

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:133]:
```suggestion
* Setting a good example for extension authors
```

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:135]:
```suggestion
* Avoiding complications and side effects during refactoring, also preventing accidental semantic versioning breaking changes in patch or feature level releases
```

Or maybe something like "prevent introducing a layer of public API responsibility"?

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:158]:
```suggestion
Utilizing named arguments in PHPUnit test Data Providers
```

(not sure, but I'd capitalize "Data Provider" as a term)

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:161]:
```suggestion
The use of named arguments in PHPUnit test Data Providers is permitted and
```

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:199]:
```suggestion
Leveraging named arguments when invoking PHP functions
```

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:202]:
```suggestion
TYPO3 Core may leverage named arguments when calling PHP functions, provided it
```

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [inline Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:204]: So no "haystack" and "needle" allowed for inconsitent PHP functions? Should this explicitly be referenced, if really not allowed? At least for me that's a key benefit when deciding to allow named parameters...?

**Review by garvinhicking — COMMENTED** [review]
Nice, I was excited to see this.

Some minor comments and questions, some siggestions mostly on not uppercasing headings anymore (hope I got this right from chris).

Exciting to fixate these practices and how it will take on in future patches.

The way I read it there won't be CI integration to mandate this for core patches, or will it be enforced for the value objects, and mass-changed once for existing code?

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (brotkrueml) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:158]: Hmm, would not capitalize it: PHPUnit itself uses it lowercased in a sentence: https://docs.phpunit.de/en/11.0/writing-tests-for-phpunit.html#data-providers

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (brotkrueml) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:202]: There are some other places left where "core" should be written "Core", see also glossar: https://docs.typo3.org/m/typo3/docs-how-to-document/main/en-us/GeneralConventions/Glossary.html

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (brotkrueml) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:6]: Yup, see here: https://docs.typo3.org/m/typo3/docs-how-to-document/main/en-us/GeneralConventions/ContentStyleGuide.html#rules-for-titles-section-headers

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (brotkrueml) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:48]: "should" is most likely used like in RFCs:
https://www.rfc-editor.org/rfc/rfc2119

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (brotkrueml) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:49]: Yup, and please use either 

    :abbr:`PCPP (public constructor property promotion)`

or write it out in brackets like `PCPP (pubic constructor property promotion`
Speaking for myself, I know CPP but not PCPP by now.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (brotkrueml) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:71]: Yup, I assume, many people don't know what value objects are.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (brotkrueml) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:108]: Haven't heard yet of "2nd-party dependency methods". What is it?

**Review by brotkrueml — COMMENTED** [review]
I will wait with a complete review, until most comments of Garvin are addressed.

Some wordings can also be streamlined after merging, like "Core" and such, by the Docs Team. Also adding abbreviations and such.

I am missing the mentioning of attributes. Currently, we use named arguments already in the docs, for example, for event listeners and commands:
https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/Events/EventDispatcher/Index.html#EventDispatcherEventListenerClass
https://docs.typo3.org/m/typo3/reference-coreapi/12.4/en-us/ApiOverview/CommandControllers/Tutorial.html#console-command-tutorial-registration-attribute

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:204]: not in method with only 2 arguments. we do *not* want to have named arguments for php functions used inflationary, just because dev can't remember implode() glue and array order. i think the rules phrasing tells that already.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:27]: Docs first, core patches to adapt ci may follow. No changelog needed.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:38]: why? it has it's own docs. and people who don't know this basic should google or something?

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:49]: feel free. it's really close below, though.

Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:49]: the explanation is just below ...

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:202]: please decide and adapt if needed.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:158]: i don't care. feel free.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:71]: feel free if you find some proper explanation, but i think people should know this ...

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:33]: no. it is the key consideration, not a drawback.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:61]: feel free to reprhase. say: "using it depends on single cases, core will try to establish CI checks for each."

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:6]: aha. feel free.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:65]: patches just for the purpose of just adding named arguments for json_decode() everywhere will be denied. we want to avoid the review overhead when the patch does not come with true advantages but just because-we-can, and cgl masturbation.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:104]: imho: keep the RFC-like outlining here, it's done in other cgl chapters, too.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:108]: core calls a non-core library?

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:114]: no. maybe later.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:27]: What I meant was, is there any PHP FIG / PSR initiative that also deals with named argument coding guidelines, which we could link to?

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:38]: We try to crosslink terms so that the documentation can be accessible for any level of developer. But Chris and I can take care of crosslinking later :-)

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:27]: did not find anything appropriate.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (garvinhicking) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:65]: How about this addition to formulate it more welcoming and positive:

"Usage of named arguments following the recommendations is welcome for any TYPO3 Core patch that introduces new functionality." ?

**Comment by lolli42** [comment]
> The way I read it there won't be CI integration to mandate this for core patches, or will it be enforced for the value objects, and mass-changed once for existing code?

"... each target of scrutinity in ci pipeline ...": we will try to add CI checks as possible and will then of course also adapt violating calls that exist.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:135]: not the intention here. 'preventing accidental semantic versioning breaking' is ruled out anyway. this is about changing a variable name of a method and then being forced to adapt all consumers as well.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:65]: unsure. 'new functionality' could be understood as 'new feature'. 'clear further benefits' of a patch matches better. i don't think this sentence needs to be positive: we want to say which patches the core team will not accept.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:71]: link to wikipedia for 'value object's if in doubt.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:106]: no. but with the existing deprecation features we have already. and we have no existing example yet.

File: Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst
Comment (lolli42) [reply Documentation/CodingGuidelines/CodingBestPractices/NamedArguments.rst:127]: yeah. please don't change this. we do not blindly create new objects with named arguments.

**Review by brotkrueml — DISMISSED** [review]

**Comment by brotkrueml** [comment]
Will streamline spellings in a follow-up.

**Review by brotkrueml — APPROVED** [review]

**Pull request merged**
